File: python/src/character.py
"""Character data model and generation logic."""

import json
import os
from copy import deepcopy
from typing import Dict, List, Optional, Any

from dice import (
    roll_abilities,
    roll_supernatural,
    characteristic_modifier,
    roll_vs_target,
    roll_2d6,
    roll_d6,
)


def load_json(path: str) -> dict:
    """Load a JSON file from the data directory."""
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def data_path(*parts: str) -> str:
    """Build a path relative to the project data directory."""
    # python/src/character.py -> project root -> data/
    base = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), "data")
    return os.path.join(base, *parts)


class Character:
    def __init__(self, name: str = "Unnamed"):
        self.name = name
        self.species_id = "human"
        self.abilities: Dict[str, int] = {}
        self.supernatural = -15
        self.hero_coins = 0
        self.background_id = ""
        self.background_skill = ""
        self.careers: List[dict] = []
        self.skills: Dict[str, int] = {}
        self.equipment: List[dict] = []
        self.assets: List[dict] = []
        self.connections: List[dict] = []
        self.age = 18
        self.total_terms = 0
        self.derived = {}

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "species_id": self.species_id,
            "abilities": self.abilities,
            "supernatural": self.supernatural,
            "hero_coins": self.hero_coins,
            "background_id": self.background_id,
            "background_skill": self.background_skill,
            "careers": self.careers,
            "skills": self.skills,
            "equipment": self.equipment,
            "assets": self.assets,
            "connections": self.connections,
            "age": self.age,
            "total_terms": self.total_terms,
            "derived": self.derived,
        }


def get_soc_tier(soc: int) -> str:
    """Return SOC tier based on exploding 2d6 result."""
    if soc <= 7:
        return "serf"
    elif soc <= 10:
        return "commoner"
    elif soc <= 13:
        return "middle"
    elif soc <= 16:
        return "gentry"
    elif soc <= 21:
        return "minor_noble"
    elif soc <= 27:
        return "major_noble"
    elif soc <= 33:
        return "high_noble"
    elif soc <= 39:
        return "minor_royalty"
    else:
        return "true_royalty"


def select_background(soc: int) -> dict:
    """Select background based on SOC roll."""
    data = load_json(data_path("backgrounds", "index.json"))
    for bg in data["backgrounds"]:
        if bg["soc_range"]["min"] <= soc <= bg["soc_range"]["max"]:
            return bg
    return data["backgrounds"][0]


def apply_species_modifiers(character: Character, species_data: dict) -> None:
    """Apply species ability modifiers to rolled abilities."""
    for stat, mod in species_data.get("ability_modifiers", {}).items():
        character.abilities[stat] = character.abilities.get(stat, 7) + mod


def award_hero_coins(character: Character) -> None:
    """Award hero coins if total ability score (excluding SOC) is low."""
    core_stats = ["str", "dex", "end", "int", "edu"]
    total = sum(character.abilities.get(s, 7) for s in core_stats)
    if total < 36:
        character.hero_coins = 3
    elif total < 42:
        character.hero_coins = 2
    elif total < 48:
        character.hero_coins = 1
    else:
        character.hero_coins = 0


def calculate_derived(character: Character) -> None:
    """Calculate derived statistics."""
    str_mod = characteristic_modifier(character.abilities.get("str", 7))
    dex_mod = characteristic_modifier(character.abilities.get("dex", 7))
    end_mod = characteristic_modifier(character.abilities.get("end", 7))
    edu_mod = characteristic_modifier(character.abilities.get("edu", 7))
    int_val = character.abilities.get("int", 7)
    edu_val = character.abilities.get("edu", 7)

    # Hit Points: END + career modifier (simplified: base END for now)
    hp = max(1, character.abilities.get("end", 7))

    # Stamina Points: END + STR modifier
    stamina = max(1, character.abilities.get("end", 7) + str_mod)

    # Initiative: DEX modifier
    initiative = dex_mod

    # Armor Class: 8 + DEX modifier (unarmored)
    ac = 8 + dex_mod

    # Skill Maximum: INT + EDU
    skill_max = int_val + edu_val

    # Languages: EDU modifier (minimum 1)
    languages = max(1, edu_mod)

    character.derived = {
        "hit_points": hp,
        "stamina_points": stamina,
        "initiative_modifier": initiative,
        "armor_class": ac,
        "skill_maximum": skill_max,
        "languages": languages,
        "movement": 6,  # Human base; updated by species
    }


def generate_base_character(name: str = "Unnamed") -> Character:
    """Generate a fresh character through background/mustering phase."""
    char = Character(name)

    # 1. Roll abilities
    char.abilities = roll_abilities()

    # 2. Load species and apply modifiers
    species_data = load_json(data_path("species", "human.json"))
    apply_species_modifiers(char, species_data)

    # 3. Roll supernatural
    char.supernatural = roll_supernatural()

    # 4. Hero coins
    award_hero_coins(char)

    # 5. SOC tier & background
    bg = select_background(char.abilities["soc"])
    char.background_id = bg["background_id"]

    # Pick one background skill
    import random
    char.background_skill = random.choice(bg["skill_options"])
    char.skills[char.background_skill] = char.skills.get(char.background_skill, 0) + 1

    # Apply species background skill bonus (Human = +1 choice)
    bonus_skills = species_data.get("background_skill_bonus", 0)
    for _ in range(bonus_skills):
        extra = random.choice(bg["skill_options"])
        char.skills[extra] = char.skills.get(extra, 0) + 1

    # 6. Starting funds
    starting_cash = int(500 * bg["starting_funds_modifier"])
    char.assets.append({"type": "cash", "amount": starting_cash})

    # 7. Derived stats
    calculate_derived(char)
    char.derived["movement"] = species_data.get("base_speed", 6)

    return char

File: src/engine/character.ts
import type { CharacterData, SpeciesData, BackgroundData } from './types'
import { rollAbilities, rollSupernatural, characteristicModifier } from './dice'
import { loadSpecies, getBackgrounds } from './dataLoader'
import { random } from './rng'

export function getSocTier(soc: number): string {
  if (soc <= 7) return 'serf'
  if (soc <= 10) return 'commoner'
  if (soc <= 13) return 'middle'
  if (soc <= 16) return 'gentry'
  if (soc <= 21) return 'minor_noble'
  if (soc <= 27) return 'major_noble'
  if (soc <= 33) return 'high_noble'
  if (soc <= 39) return 'minor_royalty'
  return 'true_royalty'
}

export function selectBackground(soc: number): BackgroundData {
  const bgs = getBackgrounds()
  for (const bg of bgs) {
    if (bg.soc_range.min <= soc && soc <= bg.soc_range.max) {
      return bg
    }
  }
  return bgs[0]
}

export function applySpeciesModifiers(character: CharacterData, speciesData: SpeciesData): void {
  for (const [stat, mod] of Object.entries(speciesData.ability_modifiers)) {
    character.abilities[stat] = (character.abilities[stat] || 7) + mod
  }
}

export function awardHeroCoins(character: CharacterData): void {
  const coreStats = ['str', 'dex', 'end', 'int', 'edu']
  const total = coreStats.reduce((sum, s) => sum + (character.abilities[s] || 7), 0)
  if (total < 36) {
    character.hero_coins = 3
  } else if (total < 42) {
    character.hero_coins = 2
  } else if (total < 48) {
    character.hero_coins = 1
  } else {
    character.hero_coins = 0
  }
}

export function calculateDerived(character: CharacterData): void {
  const strMod = characteristicModifier(character.abilities['str'] || 7)
  const dexMod = characteristicModifier(character.abilities['dex'] || 7)
  const eduMod = characteristicModifier(character.abilities['edu'] || 7)
  const intVal = character.abilities['int'] || 7
  const eduVal = character.abilities['edu'] || 7

  const hp = Math.max(1, character.abilities['end'] || 7)
  const stamina = Math.max(1, (character.abilities['end'] || 7) + strMod)
  const initiative = dexMod
  const ac = 8 + dexMod
  const skillMax = intVal + eduVal
  const languages = Math.max(1, eduMod)

  character.derived = {
    hit_points: hp,
    stamina_points: stamina,
    initiative_modifier: initiative,
    armor_class: ac,
    skill_maximum: skillMax,
    languages: languages,
    movement: 6,
  }
}

export function createCharacter(name: string = 'Unnamed'): CharacterData {
  return {
    name,
    species_id: 'human',
    abilities: {},
    supernatural: -15,
    hero_coins: 0,
    background_id: '',
    background_skill: '',
    careers: [],
    skills: {},
    equipment: [],
    assets: [],
    connections: [],
    age: 18,
    total_terms: 0,
    derived: {},
  }
}

export function generateBaseCharacter(name: string = 'Unnamed'): CharacterData {
  const char = createCharacter(name)

  // 1. Roll abilities
  char.abilities = rollAbilities()

  // 2. Load species and apply modifiers
  const speciesData = loadSpecies('human')
  applySpeciesModifiers(char, speciesData)

  // 3. Roll supernatural
  char.supernatural = rollSupernatural()

  // 4. Hero coins
  awardHeroCoins(char)

  // 5. SOC tier & background
  const bg = selectBackground(char.abilities['soc'])
  char.background_id = bg.background_id

  // Pick one background skill
  char.background_skill = bg.skill_options[Math.floor(random() * bg.skill_options.length)]
  char.skills[char.background_skill] = (char.skills[char.background_skill] || 0) + 1

  // Apply species background skill bonus (Human = +1 choice)
  const bonusSkills = speciesData.background_skill_bonus || 0
  for (let i = 0; i < bonusSkills; i++) {
    const extra = bg.skill_options[Math.floor(random() * bg.skill_options.length)]
    char.skills[extra] = (char.skills[extra] || 0) + 1
  }

  // 6. Starting funds
  const startingCash = Math.floor(500 * bg.starting_funds_modifier)
  char.assets.push({ type: 'cash', amount: startingCash })

  // 7. Derived stats
  calculateDerived(char)
  char.derived['movement'] = speciesData.base_speed

  return char
}
